rueidisprob/slidingbloomfilter.go

```diff
@@ -0,0 +1,446 @@
+package rueidisprob
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/redis/rueidis"
+)
+
+const (
+	slidingBloomFilterInitializeScript = `
+local filterKey = KEYS[1]
+local nextFilterKey = KEYS[2]
+local counterKey = KEYS[3]
+local nextCounterKey = KEYS[4]
+local lastRotationKey = KEYS[5]
+local windowHalf = tonumber(ARGV[1])
+
+if redis.call('EXISTS', filterKey, nextFilterKey, counterKey, nextCounterKey, lastRotationKey) == 0 then
+	local time = redis.call('TIME')
+	local current_time = tonumber(time[1]) * 1000 + math.floor(tonumber(time[2]) / 1000)
+
+	redis.call('MSET', filterKey, "", counterKey, 0, nextFilterKey, "", nextCounterKey, 0)
+	redis.call('SET', lastRotationKey, tostring(current_time), 'PX', windowHalf, 'NX')
+end
+
+return 1
+`
+
+	slidingBloomFilterAddMultiScript = `
+local hashIterations = tonumber(ARGV[1])
+local windowHalf = tonumber(ARGV[2])
+local numElements = tonumber(#ARGV) - 2
+
+local filterKey = KEYS[1]
+local nextFilterKey = KEYS[2]
+local counterKey = KEYS[3]
+local nextCounterKey = KEYS[4]
+local lastRotationKey = KEYS[5]
+
+local time = redis.call('TIME')
+local current_time = tonumber(time[1]) * 1000 + math.floor(tonumber(time[2])/1000)
+local acquiredLock = redis.call('SET', lastRotationKey, tostring(current_time), 'PX', windowHalf, 'NX')
+
+if acquiredLock then
+	redis.call('RENAME', nextFilterKey, filterKey)
+	redis.call('RENAME', nextCounterKey, counterKey)
+	redis.call('SET', nextFilterKey, "")
```

Comment:
Here the script always rotate the filters once if it acquires the lock. But shouldn’t it rotate the filters twice if we access it after idling for a while (> 2*windowHalf)?

Reply:
You're absolutely correct. If more than half of the window has passed since the last rotation, it means that both the current and next filters are no longer valid since they no longer contain relevant data. In this case, saving the last rotation and reseting the filter is to much time as passed will solve it